src/3pc/agent/impl.cpp

```diff
@@ -0,0 +1,679 @@
+                                     m_ticket_number.value());
+                        m_result = error_code::commit_error;
+                        do_result();
+                    },
+                    [&](const runtime_locking_shard::shard_error& e) {
+                        if(e.m_error_code
+                           == runtime_locking_shard::error_code::wounded) {
+                            m_state = state::commit_failed;
+                            m_log->trace(m_ticket_number.value(),
+                                         "wounded during commit");
+                        } else {
+                            m_state = state::commit_error;
+                            m_log->error("Shard error for commit for",
+                                         m_ticket_number.value());
+                        }
+                        m_result = error_code::commit_error;
+                        do_result();
+                    },
+                    [&](runtime_locking_shard::error_code /*e*/) {
+                        m_state = state::commit_error;
+                        m_log->error("Shard error for commit for",
+                                     m_ticket_number.value());
+                        m_result = error_code::commit_error;
+                        do_result();
+                    }},
+                res.value());
+        } else {
+            m_log->trace(this,
+                         "Agent handled commit for",
+                         m_ticket_number.value());
+            do_finish();
+        }
+    }
+
+    void impl::do_result() {
+        std::unique_lock l(m_mut);
+        assert(m_result.has_value());
+        switch(m_state) {
+            // No results should be reported in these states, fatal bugs
+            case state::init:
+                m_log->fatal("Result reported in initial state");
+            case state::begin_sent:
+                m_log->fatal("Result reported in begin_sent state");
+            case state::function_get_sent:
+                m_log->fatal("Result reported in function_get_sent state");
+            case state::commit_sent:
+                m_log->fatal("Result reported in commit_sent state");
+            case state::finish_sent:
+                m_log->fatal("Result reported in finish_sent state");
+            case state::function_started:
+                m_log->fatal("Result reported in function_started state");
+            case state::rollback_sent:
+                m_log->fatal("Result reported in rollback_sent state");
+            case state::rollback_complete:
+                if(!std::holds_alternative<error_code>(m_result.value())
+                   || std::get<error_code>(m_result.value())
+                          != error_code::retry) {
+                    m_log->fatal("Result reported in rollback_complete state "
+                                 "when result is not retry");
+                }
+                [[fallthrough]];
```

Comment:
It's curious that we're not getting the `fallthrough` warnings.

Reply:
Funny that you mention it is surprising `logger::fatal()` is able to fail. That surprises me right now, as a matter of fact. That's also on me for not having investigated whether or not it would terminate; clearly came from an untested response. Consider adding a closure task for this review item to open a bug for potentially reconsidering fatal-logging's immediate termination. Maybe it's not large enough a deal to worry about, but I'd like to not forget about it and avoid revisiting due to attrition.

Reply:
Will create an issue item for this and close.